# Cleaning data

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [15]:
df = pd.read_csv("https://raw.githubusercontent.com/su-mt4007/data/refs/heads/main/cell_phones_total.csv")

I will start with checking .info() and .describe() to learn more about the dataframe.

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 214 entries, 0 to 213
Data columns (total 57 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   iso-3   214 non-null    object 
 1   1960    200 non-null    float64
 2   1965    200 non-null    float64
 3   1966    0 non-null      float64
 4   1967    0 non-null      float64
 5   1968    0 non-null      float64
 6   1969    0 non-null      float64
 7   1970    200 non-null    float64
 8   1971    0 non-null      float64
 9   1972    0 non-null      float64
 10  1973    0 non-null      float64
 11  1974    0 non-null      float64
 12  1975    200 non-null    float64
 13  1976    200 non-null    float64
 14  1977    200 non-null    float64
 15  1978    200 non-null    float64
 16  1979    200 non-null    float64
 17  1980    200 non-null    object 
 18  1981    199 non-null    object 
 19  1982    199 non-null    object 
 20  1983    198 non-null    object 
 21  1984    196 non-null    object 
 22  19

I can see that some years only have NaN:s, this i would like to remove because it does not add any information att all.

In [7]:
df.describe()

1960   1965  1966  1967  1968  1969   1970  1971  1972  1973  1974  \
count  200.0  200.0   0.0   0.0   0.0   0.0  200.0   0.0   0.0   0.0   0.0   
mean     0.0    0.0   NaN   NaN   NaN   NaN    0.0   NaN   NaN   NaN   NaN   
std      0.0    0.0   NaN   NaN   NaN   NaN    0.0   NaN   NaN   NaN   NaN   
min      0.0    0.0   NaN   NaN   NaN   NaN    0.0   NaN   NaN   NaN   NaN   
25%      0.0    0.0   NaN   NaN   NaN   NaN    0.0   NaN   NaN   NaN   NaN   
50%      0.0    0.0   NaN   NaN   NaN   NaN    0.0   NaN   NaN   NaN   NaN   
75%      0.0    0.0   NaN   NaN   NaN   NaN    0.0   NaN   NaN   NaN   NaN   
max      0.0    0.0   NaN   NaN   NaN   NaN    0.0   NaN   NaN   NaN   NaN   

        1975   1976   1977   1978   1979  
count  200.0  200.0  200.0  200.0  200.0  
mean     0.0    0.0    0.0    0.0    0.0  
std      0.0    0.0    0.0    0.0    0.0  
min      0.0    0.0    0.0    0.0    0.0  
25%      0.0    0.0    0.0    0.0    0.0  
50%      0.0    0.0    0.0    0.0    0.0  
75%      0.0    0.0    0.0    0.0    0.0  
max      0.0    0.0    0.0    0.0    0.0

I can see that even for the years that have answers the ansers does not seem to be numerical. I will change this. When looking in the dataframe i find the letters "k", "M" and "B" representing thousend, millions and billions. I will rewrote these as given; k=1e3, M=1e6 and B=1e9. 

In [26]:
# Function to convert strings to integers
def string_to_int(value):
    try:
        if isinstance(value, str):
            if 'k' in value:  # Convert 'k' to 1e3
                return float(value.replace('k', '')) * 1e3
            elif 'M' in value:  # Convert 'M' to 1e6
                return float(value.replace('M', '')) * 1e6
            elif 'B' in value:  # Convert 'B' to 1e9
                return float(value.replace('B', '')) * 1e9
            elif value.isdigit():  # Handle plain numeric strings
                return float(value)
            else:
                # Leave unexpected strings as NaN
                return np.nan
        elif pd.isnull(value):  # Retain NaN values
            return np.nan
        else:
            return value  # Return as-is for numeric inputs
    except (ValueError, TypeError):
        # Handle any unexpected errors in conversion
        return np.nan

# Exclude "iso-3" column from conversion
df_numeric = df.copy()
for col in df_numeric.columns:
    if col != "iso-3":  # Apply conversion to all columns except "iso-3"
        df_numeric[col] = df_numeric[col].apply(string_to_int)

In [28]:
df_numeric.describe()

1960   1965  1966  1967  1968  1969   1970  1971  1972  1973  ...  \
count  200.0  200.0   0.0   0.0   0.0   0.0  200.0   0.0   0.0   0.0  ...   
mean     0.0    0.0   NaN   NaN   NaN   NaN    0.0   NaN   NaN   NaN  ...   
std      0.0    0.0   NaN   NaN   NaN   NaN    0.0   NaN   NaN   NaN  ...   
min      0.0    0.0   NaN   NaN   NaN   NaN    0.0   NaN   NaN   NaN  ...   
25%      0.0    0.0   NaN   NaN   NaN   NaN    0.0   NaN   NaN   NaN  ...   
50%      0.0    0.0   NaN   NaN   NaN   NaN    0.0   NaN   NaN   NaN  ...   
75%      0.0    0.0   NaN   NaN   NaN   NaN    0.0   NaN   NaN   NaN  ...   
max      0.0    0.0   NaN   NaN   NaN   NaN    0.0   NaN   NaN   NaN  ...   

               2010          2011          2012          2013          2014  \
count  2.070000e+02  2.030000e+02  2.070000e+02  2.040000e+02  2.050000e+02   
mean   2.545049e+07  2.886928e+07  3.009531e+07  3.251671e+07  3.402056e+07   
std    8.633080e+07  1.000190e+08  1.048954e+08  1.135204e+08  1.195511e+08   
min    1.600000e+03  2.130000e+03  2.800000e+03  3.400000e+03  3.800000e+03   
25%    6.875000e+05  1.125000e+06  1.020500e+06  1.295000e+06  1.160000e+06   
50%    4.610000e+06  5.120000e+06  5.580000e+06  6.715000e+06  7.070000e+06   
75%    1.285000e+07  1.545000e+07  1.735000e+07  1.985000e+07  2.160000e+07   
max    8.590000e+08  9.860000e+08  1.110000e+09  1.230000e+09  1.290000e+09   

               2015          2016          2017          2018          2019  
count  2.070000e+02  2.020000e+02  2.040000e+02  1.820000e+02  1.430000e+02  
mean   3.453787e+07  3.701602e+07  3.785188e+07  4.330909e+07  4.657212e+07  
std    1.214922e+08  1.324869e+08  1.396255e+08  1.568403e+08  1.766419e+08  
min    6.600000e+03  7.600000e+03  8.000000e+03  3.270000e+04  3.380000e+04  
25%    1.175000e+06  1.832500e+06  1.527500e+06  2.252500e+06  2.570000e+06  
50%    7.260000e+06  7.480000e+06  7.150000e+06  8.465000e+06  8.620000e+06  
75%    2.135000e+07  2.287500e+07  2.322500e+07  2.587500e+07  2.680000e+07  
max    1.290000e+09  1.360000e+09  1.470000e+09  1.650000e+09  1.730000e+09  

[8 rows x 56 columns]

I can now see that the dataframe have numbers and can be used the way it is suppoesed to. It does still contain some years full of only NaNs so i will remove those. I will also remove a country if there is any with no answers.

In [29]:
def remove_all_nans(df):
    # Remove columns that are all NaN
    df_cleaned = df.dropna(axis=1, how='all')
    
    # Remove rows that are all NaN
    df_cleaned = df_cleaned.dropna(axis=0, how='all')

    return df_cleaned

# Use the function on df_numeric
df_cleaned = remove_all_nans(df_numeric)

In [33]:
df_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 214 entries, 0 to 213
Data columns (total 49 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   iso-3   214 non-null    object 
 1   1960    200 non-null    float64
 2   1965    200 non-null    float64
 3   1970    200 non-null    float64
 4   1975    200 non-null    float64
 5   1976    200 non-null    float64
 6   1977    200 non-null    float64
 7   1978    200 non-null    float64
 8   1979    200 non-null    float64
 9   1980    200 non-null    float64
 10  1981    199 non-null    float64
 11  1982    199 non-null    float64
 12  1983    198 non-null    float64
 13  1984    196 non-null    float64
 14  1985    195 non-null    float64
 15  1986    196 non-null    float64
 16  1987    194 non-null    float64
 17  1988    195 non-null    float64
 18  1989    193 non-null    float64
 19  1990    207 non-null    float64
 20  1991    196 non-null    float64
 21  1992    198 non-null    float64
 22  19

Some years have been removed (no countries had only NaN:s, still 214 iso:s). We will look at it again.

In [30]:
df_cleaned.describe()

1960   1965   1970   1975   1976   1977   1978   1979          1980  \
count  200.0  200.0  200.0  200.0  200.0  200.0  200.0  200.0    200.000000   
mean     0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    117.500000   
std      0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   1661.700936   
min      0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0      0.000000   
25%      0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0      0.000000   
50%      0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0      0.000000   
75%      0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0      0.000000   
max      0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0  23500.000000   

               1981  ...          2010          2011          2012  \
count    199.000000  ...  2.070000e+02  2.030000e+02  2.070000e+02   
mean     319.949749  ...  2.545049e+07  2.886928e+07  3.009531e+07   
std     2636.532100  ...  8.633080e+07  1.000190e+08  1.048954e+08   
min        0.000000  ...  1.600000e+03  2.130000e+03  2.800000e+03   
25%        0.000000  ...  6.875000e+05  1.125000e+06  1.020500e+06   
50%        0.000000  ...  4.610000e+06  5.120000e+06  5.580000e+06   
75%        0.000000  ...  1.285000e+07  1.545000e+07  1.735000e+07   
max    28300.000000  ...  8.590000e+08  9.860000e+08  1.110000e+09   

               2013          2014          2015          2016          2017  \
count  2.040000e+02  2.050000e+02  2.070000e+02  2.020000e+02  2.040000e+02   
mean   3.251671e+07  3.402056e+07  3.453787e+07  3.701602e+07  3.785188e+07   
std    1.135204e+08  1.195511e+08  1.214922e+08  1.324869e+08  1.396255e+08   
min    3.400000e+03  3.800000e+03  6.600000e+03  7.600000e+03  8.000000e+03   
25%    1.295000e+06  1.160000e+06  1.175000e+06  1.832500e+06  1.527500e+06   
50%    6.715000e+06  7.070000e+06  7.260000e+06  7.480000e+06  7.150000e+06   
75%    1.985000e+07  2.160000e+07  2.135000e+07  2.287500e+07  2.322500e+07   
max    1.230000e+09  1.290000e+09  1.290000e+09  1.360000e+09  1.470000e+09   

               2018          2019  
count  1.820000e+02  1.430000e+02  
mean   4.330909e+07  4.657212e+07  
std    1.568403e+08  1.766419e+08  
min    3.270000e+04  3.380000e+04  
25%    2.252500e+06  2.570000e+06  
50%    8.465000e+06  8.620000e+06  
75%    2.587500e+07  2.680000e+07  
max    1.650000e+09  1.730000e+09  

[8 rows x 48 columns]

It now looks better. I guess we could remove the years with only "0" as answer but I dont really see why, "0" is still an answer and maybe you want that information to draw some kind of conclusion, so I will leave it.

Finally i will print it as in the given asignment.

In [34]:
df_numeric[['iso-3', '2015', '2016', '2017', '2018', '2019']].sort_values('2015', ascending = False).head()

iso-3          2015          2016          2017          2018  \
35    CHN  1.290000e+09  1.360000e+09  1.470000e+09  1.650000e+09   
87    IND  1.000000e+09  1.130000e+09  1.170000e+09  1.180000e+09   
200   USA  3.820000e+08  3.960000e+08  4.000000e+08  4.220000e+08   
86    IDN  3.390000e+08  3.860000e+08  4.350000e+08  3.190000e+08   
26    BRA  2.580000e+08  2.440000e+08  2.180000e+08  2.070000e+08   

             2019  
35   1.730000e+09  
87   1.150000e+09  
200           NaN  
86   3.450000e+08  
26            NaN